precompiles/gmp/src/lib.rs

```diff
@@ -0,0 +1,312 @@
+		let initial_gas = 2500 + 2 * RuntimeHelper::<Runtime>::db_read_gas_cost();
+		log::warn!("initial_gas: {:?}", initial_gas);
+		handle.record_cost(initial_gas)?;
+
+		let wormhole = storage::CoreAddress::get()
+			.ok_or(RevertReason::custom("invalid wormhole core address"))?;
+
+		let wormhole_bridge = storage::BridgeAddress::get()
+			.ok_or(RevertReason::custom("invalid wormhole bridge address"))?;
+
+		log::trace!(target: "gmp-precompile", "core contract: {:?}", wormhole);
+		log::trace!(target: "gmp-precompile", "bridge contract: {:?}", wormhole_bridge);
+
+		// get the wormhole VM from the provided VAA. Unfortunately, this forces us to parse
+		// the VAA twice -- this seems to be a restriction imposed from the Wormhole contract design
+		let output = Self::call(
+			handle,
+			wormhole,
+			EvmDataWriter::new_with_selector(PARSE_VM_SELECTOR)
+				.write(wormhole_vaa.clone())
+				.build(),
+		)?;
+		let mut reader = EvmDataReader::new(&output[..]);
+		let wormhole_vm: WormholeVM = reader.read()?;
+
+		// get the bridge transfer data from the wormhole VM payload
+		let output = Self::call(
+			handle,
+			wormhole_bridge,
+			EvmDataWriter::new_with_selector(PARSE_TRANSFER_WITH_PAYLOAD_SELECTOR)
+				.write(wormhole_vm.payload)
+				.build(),
+		)?;
+		let mut reader = EvmDataReader::new(&output[..]);
+		let transfer_with_payload: WormholeTransferWithPayloadData = reader.read()?;
+
+		// get the wrapper for this asset by calling wrappedAsset()
+		// TODO: this should only be done if needed (when token chain == our chain)
+		let output = Self::call(
+			handle,
+			wormhole_bridge,
+			EvmDataWriter::new_with_selector(WRAPPED_ASSET_SELECTOR)
+				.write(transfer_with_payload.token_chain)
+				.write(transfer_with_payload.token_address)
+				.build(),
+		)?;
+		let mut reader = EvmDataReader::new(&output[..]);
+		let wrapped_address: Address = reader.read()?;
+		log::debug!(target: "gmp-precompile", "wrapped token address: {:?}", wrapped_address);
+
+		// query our "before" balance (our being this precompile)
+		let output = Self::call(
+			handle,
+			wrapped_address.into(),
+			EvmDataWriter::new_with_selector(BALANCE_OF_SELECTOR)
+				.write(Address::from(handle.code_address()))
+				.build(),
+		)?;
+		let mut reader = EvmDataReader::new(&output[..]);
+		let before_amount: U256 = reader.read()?;
+		log::debug!(target: "gmp-precompile", "before balance: {}", before_amount);
+
+		// our inner-most payload should be a VersionedUserAction
+		let user_action = VersionedUserAction::decode_with_depth_limit(
+			32,
+			&mut transfer_with_payload.payload.as_bytes(),
+		)
+		.map_err(|_| RevertReason::Custom("Invalid GMP Payload".into()))?;
+		log::debug!(target: "gmp-precompile", "user action: {:?}", user_action);
+
+		let currency_account_id = Runtime::AddressMapping::into_account_id(wrapped_address.into());
+
+		let currency_id: <Runtime as orml_xtokens::Config>::CurrencyId =
```

Comment:
Technically, the wrapped address will be always a real ERC20 contract? If it's the case, you can convert it directly to the good variant.
I even think that it is deeply inefficient to go through xtokens here, we should send an XCM message directly with a generic `XcmRouter`, we can do this because from XCM's point of view, the reserve chain will always be moonbeam.

Reply:
I added an internal issue to implement these optimizations.